src/reactors/V3DutchOrderReactor.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+import {BaseReactor} from "./BaseReactor.sol";
+import {IPermit2} from "permit2/src/interfaces/IPermit2.sol";
+import {Permit2Lib} from "../lib/Permit2Lib.sol";
+import {ExclusivityLib} from "../lib/ExclusivityLib.sol";
+import {V3DutchDecayLib} from "../lib/V3DutchDecayLib.sol";
+import {
+    V3DutchOrderLib,
+    V3DutchOrder,
+    CosignerData,
+    V3DutchOutput,
+    V3DutchInput
+} from "../lib/V3DutchOrderLib.sol";
+import {SignedOrder, ResolvedOrder} from "../base/ReactorStructs.sol";
+
+/// @notice Reactor for non-linear dutch orders
+/// @dev Non-linear orders must be cosigned by the specified cosigner to override starting block and value
+/// @dev resolution behavior:
+/// - If cosignature is invalid or not from specified cosigner, revert
+/// - If inputAmount is 0, then use baseInput
+/// - If inputAmount is nonzero, then ensure it is less than specified baseInput and replace startAmount
+/// - For each outputAmount:
+///   - If amount is 0, then use baseOutput
+///   - If amount is nonzero, then ensure it is greater than specified baseOutput and replace startAmount
+contract V3DutchOrderReactor is BaseReactor {
+    using Permit2Lib for ResolvedOrder;
+    using V3DutchOrderLib for V3DutchOrder;
+    using V3DutchDecayLib for V3DutchOutput[];
+    using V3DutchDecayLib for V3DutchInput;
+    using ExclusivityLib for ResolvedOrder;
+
+    /// @notice thrown when an order's deadline is passed
+    error DeadlineReached();
+
+    /// @notice thrown when an order's cosignature does not match the expected cosigner
+    error InvalidCosignature();
+
+    /// @notice thrown when an order's cosigner input is greater than the specified
+    error InvalidCosignerInput();
+
+    /// @notice thrown when an order's cosigner output is less than the specified
+    error InvalidCosignerOutput();
+
+    constructor(IPermit2 _permit2, address _protocolFeeOwner) BaseReactor(_permit2, _protocolFeeOwner) {}
+
+    /// @inheritdoc BaseReactor
+    function _resolve(SignedOrder calldata signedOrder)
+        internal
+        view
+        virtual
+        override
+        returns (ResolvedOrder memory resolvedOrder)
+    {
+        V3DutchOrder memory order = abi.decode(signedOrder.order, (V3DutchOrder));
+        // hash the order _before_ overriding amounts, as this is the hash the user would have signed
+        bytes32 orderHash = order.hash();
+
+        if (order.info.deadline < block.timestamp) {
```

Comment:
can move this into validation fn?

Reply:
Yeah, I just need to make it a `view` but it works.